### PR TITLE
implement specification of #6

### PR DESCRIPTION
**Pull request by jklmnn**
I refined the implementation to the specs discussed in #6. `doc/exceptions.md` contains the exception type table in markdown which defines the exception names and values. `tools/generate_exceptions.py` is a script which takes the markdown file as argument and puts `ada_exceptions.h` and `ada_exceptions.ads` into the current directory (so run it from `src/`). This is just to regenerate them when something changed and shouldn't be done in the build process. I provided the generated files also in the source.
Please review and check if I forgot anything.

### CODE REVIEW

File: src/ada_exceptions.ads [hunk 1]
```diff
@@ -0,0 +1,85 @@
...
+      PE_Potentially_Blocking_Operation => 16#20e#,
+      PE_Stream_Operation_Not_Allowed => 16#20f#,
+      PE_Stubbed_Subprogram_Called => 16#210#,
+      PE_Unchecked_Union_Restriction => 16#211#,
+      PE_Finalize_Raised_Exception => 16#212#,
+      SE_Explicit_Raise => 16#300#,
+      SE_Empty_Storage_Pool => 16#301#,
+      SE_Infinite_Recursion => 16#302#,
+      SE_Object_Too_Large => 16#303#
+    );
```
Comment (treiher) [inline src/ada_exceptions.ads]: add newline

File: src/ada_exceptions.ads [hunk 2]
```diff
@@ -0,0 +1,85 @@
...
+end Ada_Exceptions;
```
Comment (treiher) [inline src/ada_exceptions.ads]: add newline

File: src/ada_exceptions.h [hunk 3]
```diff
@@ -0,0 +1,40 @@
...
+    PE_POTENTIALLY_BLOCKING_OPERATION = 0x20e,
+    PE_STREAM_OPERATION_NOT_ALLOWED = 0x20f,
+    PE_STUBBED_SUBPROGRAM_CALLED = 0x210,
+    PE_UNCHECKED_UNION_RESTRICTION = 0x211,
+    PE_FINALIZE_RAISED_EXCEPTION = 0x212,
+    SE_EXPLICIT_RAISE = 0x300,
+    SE_EMPTY_STORAGE_POOL = 0x301,
+    SE_INFINITE_RECURSION = 0x302,
+    SE_OBJECT_TOO_LARGE = 0x303
+} exception_t;
```
Comment (treiher) [inline src/ada_exceptions.h]: add newline

File: src/platform.ads [hunk 4]
```diff
@@ -0,0 +1,80 @@
...
+      Convention => C,
+      External_Name => "__gnat_free";
+
+   procedure Unhandled_Terminate
+      with
+      Import,
+      Convention => C,
+      External_Name => "__gnat_unhandled_terminate";
+
+   function Allocate_Secondary_Stack (
```
Comment (treiher) [inline src/platform.ads]: fix style

File: src/platform.ads [hunk 5]
```diff
@@ -0,0 +1,80 @@
...
+      External_Name => "allocate_secondary_stack";
+
+   function Get_Thread return System.Address
+      with
+      Import,
+      Convention => C,
+      External_Name => "get_thread";
+
+   procedure Raise_Ada_Exception (T : Ada_Exceptions.Exception_Type;
+                           Name : String;
```
Comment (treiher) [inline src/platform.ads]: fix indention

File: tools/generate_exceptions.py [hunk 6]
```diff
@@ -0,0 +1,42 @@
+#!/usr/bin/env python3
+
+import sys
+
+exceptions = {}
+
+def gen_header():
+    return "\
```
Comment (treiher) [inline tools/generate_exceptions.py]: use """ for block string instead of escaping each line

File: tools/generate_exceptions.py [hunk 7]
```diff
@@ -0,0 +1,42 @@
...
+" + "\n".join(["      {} => 16#{}#,".format(exceptions[name], hex(name)[2:]) for name in sorted(list(exceptions))]).strip(",") + "\n\
+    );\n\
+end Ada_Exceptions;"
+
+with open(sys.argv[1], "r") as md:
+    for line in md.readlines():
+        try:
+            exception = line.split("|")[1].strip(" `")
+            excid = line.split("|")[2].strip(" `")
+            if not exception[1] == 'E' and not excid.startswith("0x"):
```
Comment (treiher) [inline tools/generate_exceptions.py]: fix condition

File: Makefile [hunk 8]
```diff
@@ -2,7 +2,7 @@ COMPONENT = rts
 OBJ_DIR = obj
 TEST_DIR = tests
 
-SRC = a-except.adb a-unccon.ads ada.ads i-c.adb i-cexten.ads i-cstrin.adb interfac.ads external.adb ss_utils.adb s-exctab.adb s-imgint.adb s-init.adb s-parame.ads s-secsta.adb s-soflin.adb s-stalib.ads s-stoele.adb system.ads argv.c exit.c init.c
+SRC = a-except.adb a-unccon.ads ada.ads i-c.adb i-cexten.ads i-cstrin.adb interfac.ads platform.adb ss_utils.adb s-exctab.adb s-imgint.adb s-init.adb s-parame.ads s-secsta.adb s-soflin.adb s-stalib.ads s-stoele.adb system.ads argv.c exit.c init.c ada_exceptions.ads s-unstyp.ads
```
Comment (treiher) [inline Makefile]: sort alphabetically (you could also convert it into a list)

File: src/a-except.adb [hunk 9]
```diff
@@ -41,18 +42,16 @@ package body Ada.Exceptions is
...
-   begin
-      External.Warn_Not_Implemented ("Raise_Exception_Always");
-      Raise_Ada_Exception (E.Full_Name, C_Msg'Address);
+      pragma Unreferenced (E);
+   begin
+      Platform.Log_Warning  ("Raise_Exception_Alwaysis not implemented");
+      Platform.Raise_Ada_Exception (Ada_Exceptions.
+                                       Undefined_Exception,
+                                    "Undefined_Exception",
+   --  It is not sure that E.Full_Name is not null
```
Comment (treiher) [inline src/a-except.adb]: Does that comment help at this position?

File: src/platform.ads [hunk 10]
```diff
@@ -0,0 +1,80 @@
...
+package Platform is
+   pragma Preelaborate;
+
+   procedure Log_Debug (Msg : String);
+   procedure Log_Warning (Msg : String);
+   procedure Log_Error (Msg : String);
+
+   function Malloc (Size : Natural) return System.Address
+      with
+      Import,
```
Comment (treiher) [inline src/platform.ads]: fix indention

**Review by treiher — CHANGES_REQUESTED** [review]

**Comment by jklmnn** [comment]
https://github.com/Componolit/ada-runtime/pull/7/commits/4cd045a2fc2ab3c92ee933c5fc8931eb4ecdf96d should fix the comments you added. I linted the Ada code with GPS so it is correct by definition ;)

File: src/a-except.adb [hunk 11]
```diff
@@ -161,556 +160,556 @@ package body Ada.Exceptions is
 
    procedure Rcheck_CE_Access_Check (File : System.Address;
                                      Line : Integer)
-      with Export,
-           Convention => C,
-           External_Name => "__gnat_rcheck_CE_Access_Check";
+     with Export,
```
Comment (treiher) [inline src/a-except.adb]: Could you please make the style of the aspect consistent with `platform.ads`?

```
procedure X
  with
    Export,
    Convention ...
```

Comment (jklmnn) [reply src/a-except.adb]: Fixed by https://github.com/Componolit/ada-runtime/pull/7/commits/f7181fb1083065b64927b930c745d5620c1d21d9.

**Comment by treiher** [comment]
Merged: https://github.com/Componolit/ada-runtime/commit/d1e649fee1c24016b364848cd1f56591e74bafcb

**Pull request closed without merging**